cmd/snap-confine/seccomp-support.c

```diff
@@ -109,12 +136,73 @@ static void validate_bpfpath_is_safe(const char *path)
 	}
 }
 
+static void sc_cleanup_sock_fprog(struct sock_fprog *prog)
+{
+	free(prog->filter);
+	prog->filter = NULL;
+}
+
+static void sc_must_read_filter_from_file(FILE *file, uint32_t len_bytes, char *what, struct sock_fprog *prog)
+{
+	prog->len = len_bytes / sizeof(struct sock_filter);
+	prog->filter = (struct sock_filter *)malloc(len_bytes);
+	if (prog->filter == NULL) {
+		die("cannot allocate %u bytes of memory for %s seccomp filter ", len_bytes, what);
+	}
+	size_t num_read = fread(prog->filter, 1, len_bytes, file);
+	if (ferror(file)) {
+		die("cannot read %s filter", what);
+	}
+	if (num_read != len_bytes) {
+		die("short read for filter %s %zu != %i", what, num_read, len_bytes);
+	}
+}
+
+static FILE* sc_must_read_header_from_file(const char *profile_path, struct sc_seccomp_file_header *hdr)
+{
```

Comment:
Yeah, mmap() is probably better. I wonder if this should be a followup though because it will require some more churn and we have not used mmap() at all in snap-confine yet (it should not raise new security concerns but given that snap-confine is suid root we need to be careful). The sizes should all be pretty small (a bpf program is currently limited to 32kb) so it should be okay for now.

Reply:
That is fine. I was only thinking mmap would make the code simpler.